REBASE.openshift.md

```diff
@@ -0,0 +1,270 @@
+# Maintaining openshift/kubernetes
+
+OpenShift is based on upstream Kubernetes. With every release of Kubernetes that is
+intended to be shipped as OCP, it is necessary to incorporate the upstream changes
+while ensuring that our downstream customizations are maintained.
+
+## Rebasing for releases < 4.6
+
+The instructions in this document apply to OpenShift releases 4.6 and
+above. For previous releases, please see the [rebase
+enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/rebase.md).
+
+## Getting started
+
+Before incorporating upstream changes you may want to:
+
+- Read this document
+- Get familiar with tig (text-mode interface for git)
+- Find the best tool for resolving merge conflicts
+- Use diff3 conflict resolution strategy
+   (https://blog.nilbus.com/take-the-pain-out-of-git-conflict-resolution-use-diff3/)
+- Teach Git to remember how you’ve resolved a conflict so that the next time it can
+  resolve it automatically (https://git-scm.com/book/en/v2/Git-Tools-Rerere)
+
+## Preparing the local repo clone
+
+Clone from a personal fork of kubernetes:
+
+```
+git clone https://github.com/<user id>/kubernetes
+```
+
+Enable push for personal fork:
+
+```
+git remote set-url --push origin git@github.com:<user id>/kubernetes.git
+```
+
+Add a remote for upstream and fetch its branches:
+
+```
+git remote add --fetch upstream https://github.com/kubernetes/kubernetes
+```
+
+Add a remote for the openshift fork and fetch its branches:
+
+```
+git remote add --fetch openshift https://github.com/openshift/kubernetes
+```
+
+## Creating a new local branch for the new rebase
+
+- Branch the target `k8s.io/kubernetes` release tag (e.g. `v1.20.0`) to a new
+  local branch
+
+```
+git checkout -b rebase-1.20.0 v1.20.0
+```
+
+- Merge the targeted `openshift/kubernetes` branch (e.g. `master`) with
+  strategy `ours` to reset the the branch to the targeted release tag without
+  involving manual conflict resolution.
+
+```
+git merge -s ours openshift/master
```

Comment:
My initial though is to do a rebase here, solve the conflicts and then compare the patches with the spreadsheet to be sure. I'll get back here with feedback when I actually try it.

Reply:
yeah, I would have thought the same.

Reply:
I think we need to merge or we create different upstream commits in the history that won't match on the next rebase

Reply:
but shouldn't we merge the other way? at some point we've merged k-master into ocp-master, so we want to start from ocp-master and merge updated k-master into it to just get new commit from v1.19.0..v1.19.2 where v1.19.0 is the common merge base

Reply:
namely:
```
git fetch upstream --tags

git checkout -b ocp-rebase-1.19.2 openshift/master --no-track

git merge v1.19.2
```

Reply:
scratch that, we need patches to be applied as the latest, I got lost there.

Reply:
Yeah, you've rediscovered my experience from experimenting with this as part of the transition. Using strategy 'ours' is the only reliable way to ensure that we're able to pick the carry commits cleanly. Anything else involves manual conflict resolution that risks introducing problems for no real benefit.